11_FFTBloom/app_resources/fft_convolve_ifft.hlsl

```diff
@@ -0,0 +1,205 @@
+/* [[vk::combinedImageSampler]]*/ [[vk::binding(0, 0)]] SamplerState samplerState[CHANNELS] : register(s0);
+
+#define FFT_LENGTH (_NBL_HLSL_WORKGROUP_SIZE_ * ELEMENTS_PER_THREAD)
+
+groupshared uint32_t sharedmem[workgroup::fft::SharedMemoryDWORDs<scalar_t, _NBL_HLSL_WORKGROUP_SIZE_>];
+
+// Users MUST define this method for FFT to work
+uint32_t3 glsl::gl_WorkGroupSize() { return uint32_t3(_NBL_HLSL_WORKGROUP_SIZE_, 1, 1); }
+
+struct SharedMemoryAccessor
+{
+	void set(uint32_t idx, uint32_t value)
+	{
+		sharedmem[idx] = value;
+	}
+
+	void get(uint32_t idx, NBL_REF_ARG(uint32_t) value)
+	{
+		value = sharedmem[idx];
+	}
+
+	void workgroupExecutionAndMemoryBarrier()
+	{
+		glsl::barrier();
+	}
+
+};
+
+// ---------------------------------------------------- Utils ---------------------------------------------------------
+uint64_t colMajorOffset(uint32_t x, uint32_t y)
+{
+	return x * glsl::gl_NumWorkGroups().x | y; // can sum with | here because NumWorkGroups is still PoT (has to match half the FFT_LENGTH of previous pass)
+}
+
+uint64_t rowMajorOffset(uint32_t x, uint32_t y)
+{
+	return y * pushConstants.dataElementCount + x; // can no longer sum with | since there's no guarantees on row length
+}
+
+// Same as what was used to store in col-major after first axis FFT. This time we launch one workgroup per row so the height of the channel's (half) image is `glsl::gl_NumWorkGroups().x`,
+// and the width (number of columns) is passed as a push constant
+uint64_t getColMajorChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.colMajorBufferAddress + channel * glsl::gl_NumWorkGroups().x * pushConstants.dataElementCount * sizeof(complex_t<scalar_t>);
+}
+
+// Image saved has the same size as image read 
+uint64_t getRowMajorChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.rowMajorBufferAddress + channel * glsl::gl_NumWorkGroups().x * pushConstants.dataElementCount * sizeof(complex_t<scalar_t>);
+}
+
+
+// ------------------------------------------ SECOND AXIS FFT + CONVOLUTION + IFFT -------------------------------------------------------------
+
+// This time each Workgroup will compute the FFT along a horizontal line (fixed y for the whole Workgroup). We get the y coordinate for the
+// row a workgroup is working on via `gl_WorkGroupID().x`. We have to keep this in mind: What's stored as the first row is actually `Z + iN`, 
+// where `Z` is the actual 0th row and `N` is the Nyquist row (the one with index FFT_LENGTH / 2). Those are packed together
+// so they need to be unpacked properly after FFT like we did earlier.
+
+struct PreloadedSecondAxisAccessor : PreloadedAccessorBase<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_, scalar_t>
+{
+	int32_t mirrorWrap(int32_t paddedCoordinate)
+	{
+		const int32_t negMask = paddedCoordinate >> 31u;
+		const int32_t d = ((paddedCoordinate ^ negMask) / pushConstants.dataElementCount) ^ negMask;
```

Comment:
integer division is horribly slow on GPUs, think of something else to implement this wrap mode (or do clamp to edge instead

Reply:
Would passing 1/ dataElementCount as a float then do round of the product work?

Reply:
mirror can be done a lot more effectively
```cpp
uint32_t mirrored = unpadded;
if (unpadded<padding)
   padded = (elementCount+padding)-unpadded;
else if (unpadded>elementCountPlusOnePadding)
   padded = (2*elementCount-1)-unpadded;
```
the bits in parentheses can be precomputed as push constants

Reply:
I'm not following the math in this one